vpr/src/route/route_path_manager.cpp

```diff
@@ -0,0 +1,128 @@
+#include "route_path_manager.h"
+#include "globals.h"
+
+PathManager::PathManager(bool init_rcv_structures) {
+    // Only init data structure if required by RCV
+    // TODO: Use arena allocation method in the future, which would be setup here
+
+    if (init_rcv_structures) {
+        _is_enabled = true;
+    } else {
+        _is_enabled = false;
+    }
+}
+
+PathManager::~PathManager() {
+    if (_is_enabled) free_all_memory();
+}
+
+bool PathManager::node_exists_in_tree(t_heap_path* path_data,
+                                      RRNodeId& to_node) {
+    // Prevent seg faults for searching path data structures that haven't been created yet
+    if (!path_data || !_is_enabled) return false;
+
+    // First check the smaller current path, the ordering of these checks might effect runtime slightly
+    for (auto& node : path_data->path_rr) {
+        if (node == to_node) {
+            return true;
+        }
+    }
+
+    // Search through route tree set for nodes existance
+    auto node_exists_in_route_tree = route_tree_nodes.find(to_node);
+
+    if (node_exists_in_route_tree != route_tree_nodes.end()) {
+        return true;
+    }
+
+    return false;
+}
+
+void PathManager::insert_node(RRNodeId node) {
+    if (_is_enabled) {
+        route_tree_nodes.insert(node);
+    }
+}
+
+void PathManager::insert_backwards_path_into_traceback(t_heap_path* path_data, float cost, float backward_path_cost) {
```

Comment:
Rather than grab the route context from the g_vpr_ctx, pass in the the array view from `ConnectionRouter`.

Reply:
Still outstanding